File: rag-langchain-python/vector_query_retriever.py
from typing import List
from langchain_core.documents import Document
from langchain_core.retrievers import BaseRetriever
from langchain_openai import AzureOpenAIEmbeddings
from marklogic import Client


class VectorQueryRetriever(BaseRetriever):
    client: Client
    embedding_generator: AzureOpenAIEmbeddings

    @classmethod
    def create(
        cls,
        client: Client,
        embedding_generator: AzureOpenAIEmbeddings
    ):
        return cls(
            client=client,
            embedding_generator=embedding_generator
        )

    def _build_eval_script(self, query, query_embedding):
        words = []
        for word in query.split():
            if len(word) > 2:
                words.append(word.lower().replace("?", ""))

        return """
const op = require('/MarkLogic/optic');

op.fromSearchDocs(
    cts.andQuery([cts.wordQuery({}), cts.collectionQuery('events')]),
    null,
    {{
      'scoreMethod': 'score-bm25',
      'bm25LengthWeight': 0.5
    }}
  )
  .limit(100)
  .bind(op.as('transcript', op.xpath('doc', '/transcript')))
  .joinInner(
    op.fromView('example','events', '', op.fragmentIdCol('vectorsDocId')),
    op.on(
      op.fragmentIdCol('fragmentId'),
      op.fragmentIdCol('vectorsDocId')
    )
  )
  .bind(op.as('cosineSim',
    op.vec.cosineSimilarity(
        op.vec.vector(op.col('embedding')),
        op.vec.vector(vec.vector({}))
    )
  ))
  .bind(op.as('hybridScore',
    op.vec.vectorScore(op.col('score'), op.col('cosineSim'), 0.1)
  ))
  .select(['uri', 'transcript', 'hybridScore'])
  .orderBy(op.desc(op.col('hybridScore')))
  .limit(10)
  .result()
        """.format(
            words,
            query_embedding
        )

    def _get_relevant_documents(self, query: str) -> List[Document]:
        query_embedding = self.embedding_generator.embed_query(query)
        eval_script = self._build_eval_script(query, query_embedding)
        optic_rows = self.client.eval(javascript=eval_script)
        print(optic_rows[1].keys())

        print(f"Count of MarkLogic chunks sent to the LLM: {len(optic_rows)}")
        for optic_row in optic_rows:
            print(f"URI: {optic_row['uri']}")
        return map(lambda optic_row: Document(page_content=optic_row["transcript"]), optic_rows)

File: rag-langchain-js/vectorQueryRetriever.js
import { AzureOpenAIEmbeddings } from "@langchain/openai";
import { BaseRetriever } from "@langchain/core/retrievers";
import { Document } from "@langchain/core/documents";
import {config} from "dotenv";
import pkg from 'marklogic';
const { ctsQueryBuilder, planBuilder } = pkg;

export class VectorQueryRetriever extends BaseRetriever {
  lc_namespace = ["langchain", "retrievers"];
  marklogicClient;
  embeddingModel;

  constructor(fields) {
    super(fields);
    this.marklogicClient = fields.marklogicClient;
    config({path: "../.env"});
    this.embeddingModel = new AzureOpenAIEmbeddings({
      azureOpenAIApiDeploymentName: process.env.AZURE_EMBEDDING_DEPLOYMENT_NAME,
    });
  }

  #buildWordsArray(chatQuestion) {
    const words = [];
    chatQuestion.split(" ").forEach(word => {
      if (word.length > 2) words.push('"' + word.toLowerCase().replace("?", "") + '"')
    })
    return words;
  }

  async #generateEmbedding(chatQuestion) {
    return this.embeddingModel.embedQuery(chatQuestion);
  }

  #buildOpticPlan(words, chatQuestionEmbedding) {
    return planBuilder
      .fromSearchDocs(ctsQueryBuilder.cts.andQuery([
        ctsQueryBuilder.cts.wordQuery(words),
        ctsQueryBuilder.cts.collectionQuery("events")
      ]))
      .limit(100)
      .joinInner(
        planBuilder.fromView("example", "events", "", planBuilder.fragmentIdCol("vectorsDocId")),
        planBuilder.on(planBuilder.fragmentIdCol("fragmentId"), planBuilder.fragmentIdCol("vectorsDocId"))
      )
      .bind(planBuilder.as(planBuilder.col("similarity"), planBuilder.vec.cosineSimilarity(
        planBuilder.vec.vector(planBuilder.col("embedding")),
        planBuilder.vec.vector(chatQuestionEmbedding)
      )))
      .orderBy(planBuilder.desc(planBuilder.col("similarity")))
      .select("uri", "text")
      .limit(10);
  }

  async _getRelevantDocuments(
    chatQuestion, _callbacks,
  ) {
    const words = this.#buildWordsArray(chatQuestion);
    const chatQuestionEmbedding = await this.#generateEmbedding(chatQuestion);

    const opticQuery = this.#buildOpticPlan(words, chatQuestionEmbedding);
    const opticQueryResults = await this.marklogicClient.documents.query(opticQuery).result((results) => {
      return results;
    });

    const documents = [];
    opticQueryResults.forEach((result) => {
      console.log(result["uri"])
      documents.push(
        new Document({
          pageContent: result["content"]["transcript"],
          metadata: {"uri": result["uri"]}
        })
      )
    });
    return documents;
  }
}
